### PR TITLE
Use macOS 14 in CI

### PR DESCRIPTION
**Pull request by ISSOtm**
So far `make` works, `cmake` doesn't.

### CODE REVIEW

**Comment by Rangi42** [comment]
Oof, macOS is already the slowest CI to build, and this doubles them.

File: .github/workflows/testing.yml [hunk 1]
```diff
@@ -79,8 +81,7 @@ jobs:
   macos-static:
     strategy:
       matrix:
-        # Don't run on macOS 11; our setup makes clang segfault (YES).
-        os: [macos-12]
+        os: [macos-12, macos-14]
       fail-fast: false
     runs-on: ${{ matrix.os }}
```
Comment (ISSOtm) [inline .github/workflows/testing.yml:86]: We actually only need a single OS for this (the same one as the release workflow), so actually, using the `matrix` is overkill.

```yaml
    runs-on: macos-14
```

(That workflow should be updated accordingly, too.)

**Comment by ISSOtm** [comment]
There is only one remaining failure, and it's building with CMake. (The rest turned out to be errors on our side, not all of which directly stem from this PR's changes!)

Compare Make and CMake:
```diff
-clang++ -ggdb3 -Og -fno-omit-frame-pointer -fno-optimize-sibling-calls -Wall -pedantic -Wno-unknown-warning-option -Wno-gnu-zero-variadic-macro-arguments -Werror -Wextra 	-Walloc-zero -Wcast-align -Wcast-qual -Wduplicated-branches -Wduplicated-cond 	-Wfloat-equal -Wlogical-op -Wnull-dereference -Wshift-overflow=2 	-Wstringop-overflow=4 -Wstrict-overflow=5 -Wundef -Wuninitialized -Wunused 	-Wshadow 	-Wformat=2 -Wformat-overflow=2 -Wformat-truncation=1 	-Wno-format-nonliteral -Wno-strict-overflow -Wno-unused-but-set-variable 	-Wno-type-limits -Wno-tautological-constant-out-of-range-compare 	-Wvla 	-D_GLIBCXX_ASSERTIONS 	-fsanitize=shift -fsanitize=integer-divide-by-zero 	-fsanitize=unreachable -fsanitize=vla-bound 	-fsanitize=signed-integer-overflow -fsanitize=bounds 	-fsanitize=object-size -fsanitize=bool -fsanitize=enum 	-fsanitize=alignment -fsanitize=null -fsanitize=address -std=c++2a -I include -D_POSIX_C_SOURCE=200809L -fno-exceptions -fno-rtti -c -o src/asm/parser.o src/asm/parser.cpp
+/usr/bin/clang++ -D_GLIBCXX_ASSERTIONS -D_POSIX_C_SOURCE=200809L -I/Users/runner/work/rgbds/rgbds/include -g -Og -fno-omit-frame-pointer -fno-optimize-sibling-calls -std=gnu++20 -arch arm64 -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk -Wall -pedantic -Wno-gnu-zero-variadic-macro-arguments -fsanitize=shift -fsanitize=integer-divide-by-zero -fsanitize=unreachable -fsanitize=vla-bound -fsanitize=signed-integer-overflow -fsanitize=bounds -fsanitize=object-size -fsanitize=bool -fsanitize=enum -fsanitize=alignment -fsanitize=null -fsanitize=address -Werror -Wextra -Walloc-zero -Wcast-align -Wcast-qual -Wduplicated-branches -Wduplicated-cond -Wfloat-equal -Wlogical-op -Wnull-dereference -Wshift-overflow=2 -Wstringop-overflow=4 -Wstrict-overflow=5 -Wundef -Wuninitialized -Wunused -Wshadow -Wformat=2 -Wformat-overflow=2 -Wformat-truncation=1 -Wno-format-nonliteral -Wno-type-limits -Wno-tautological-constant-out-of-range-compare -Wvla -Wno-unknown-warning-option -MD -MT src/CMakeFiles/rgbasm.dir/asm/parser.cpp.o -MF CMakeFiles/rgbasm.dir/asm/parser.cpp.o.d -o CMakeFiles/rgbasm.dir/asm/parser.cpp.o -c /Users/runner/work/rgbds/rgbds/src/asm/parser.cpp
```

Removing flags that are strictly identical between the two (order notwithstanding):

```diff
-clang++ -ggdb3 -Wno-strict-overflow -Wno-unused-but-set-variable 	-std=c++2a -I include -fno-exceptions -fno-rtti -c -o src/asm/parser.o src/asm/parser.cpp
+/usr/bin/clang++ -I/Users/runner/work/rgbds/rgbds/include -g -std=gnu++20 -arch arm64 -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk -MD -MT src/asm/parser.cpp.o -MF asm/parser.cpp.o.d -o asm/parser.cpp.o -c /Users/runner/work/rgbds/rgbds/src/asm/parser.cpp
```

Removing flags that (**should be**) semantically equivalent:

Make | CMake
-|-
`-ggdb3` | `-g`
`-c -o ....o ....cpp` | Same but different paths
`-I include` | `-I/Users/runner/work/rgbds/rgbds/include`
`-std=c++2a` | `-std=gnu++20`

...leaves us with:

```diff
-clang++ -Wno-strict-overflow -Wno-unused-but-set-variable 	-fno-exceptions -fno-rtti
+/usr/bin/clang++ -arch arm64 -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk -MD -MT src/asm/parser.cpp.o -MF asm/parser.cpp.o.d
```

...and at this point I'm not sure what could be causing the difference.

**Comment by Rangi42** [comment]
`clang++` needs `-Wno-unused-but-set-variable`; `make` specifies that but not `cmake`.

**Review by Rangi42 — APPROVED** [review]

**Pull request merged**